packages/flame/lib/src/components/mixins/has_paint.dart

```diff
@@ -97,4 +98,32 @@ mixin HasPaint<T extends Object> on Component {
   void tint(Color color, {T? paintId}) {
     getPaint(paintId).colorFilter = ColorFilter.mode(color, BlendMode.srcATop);
   }
+
+  @override
+  double get opacity => paint.color.opacity;
+
+  @override
+  set opacity(double value) {
+    paint.color = paint.color.withOpacity(value);
+    for (final paint in _paints.values) {
+      paint.color = paint.color.withOpacity(value);
+    }
+  }
+
+  OpacityProvider opacityProviderForPaint(T paintId) {
+    return _ProxyOpacityProvider(paintId, this);
+  }
+}
```

Comment:
How about adding a provider for all opacities at once? This seems to be a common case, where you have a multi-paint component, and want to fade it out as a whole.

Reply:
> This seems to be a common case, where you have a multi-paint component, and want to fade it out as a whole.

This use case is already handled by `opacity` setter of `HasPaint`. Even if we add a provider for all paints, how do we handle `opacity` getter if all paints do not have the same `opacity`?

Reply:
I would say that if you have a class with 2 paints, one with opacity 90% and another at 25%, then `get`ting the overall opacity should return 90% (the most opaque component), and `set`ting it to, say 50%`, should make the first paint go to 50% and the second paint to (25/90*50)% -- that is, the ratio of the opacities should remain the same.

This means the opacity provider has to store these ratios at the time of construction, and then use them later on.

Reply:
Isn't it overkill to have logic for that, the user can just handle the colors separately if they need a set ratio between the opacities?

It also gets a bit weird since the colors with less opacity never can get above what they started at if another color started on 1.0.

Reply:
> Isn't it overkill to have logic for that in here?

Not an overkill, since the logic will be in a separate `OpacityProvider` class.

> It also gets a bit weird since the colors with less opacity never can get above what they started at if another color started on 1.0.

Why would that be a problem? If you draw something with a faint color on top of a solid color -- for example, a watermark on a page -- then you'd want it to keep roughly the same opacity ratio relative to the whole component. On the other hand, if you wanted to make the water mark more or less visible -- then the OpacityEffect can be applied to that individual paint.

Reply:
> Why would that be a problem? If you draw something with a faint color on top of a solid color -- for example, a watermark on a page -- then you'd want it to keep roughly the same opacity ratio relative to the whole component. On the other hand, if you wanted to make the water mark more or less visible -- then the OpacityEffect can be applied to that individual paint.

Sure, but that is only one usecase, I'm sure there are others.

> then the OpacityEffect can be applied to that individual paint.

True, that would get a very strange behavior if it wasn't done in this manner, I'm convinced.

Reply:
> for example, a watermark on a page -- then you'd want it to keep roughly the same opacity ratio relative to the whole component

So if such a component is completely faded out, both paint opacities will become zero. After this, if the component is faded in using a new `OpacityProvider`, it will not maintain the original ratios.

> `get`ting the overall opacity should return 90% (the most opaque component)

This approach doesn't really define an absolute true opacity. We are returning the opacity of most **opaque** paint and that can change anytime depending on how other opacity effects modify the paints. Even paints can be added and removed by the users anytime. I think this will involve a lot of book keeping to maintain the ratios.

Reply:
Surely there could be some corner cases when the effect doesn't work quite right. And if someone encounters such case in real life, they will have to find a way to deal with that problem (and perhaps even submit a PR).

However, from a practical standpoint, the most common scenario is that when there is a component with several paints (such as shape with background and border, and perhaps some text), then it is useful to be able to fade out (or fade in) that component as one whole, which involves maintaining ratios between opacities of individual paints.

Reply:
Added a `_MultiPaintOpacityProvider` which allows creation of a provider based on a list of paint ids